### ISSUE TITLE
New item

### ISSUE THREAD

**Issue by rosiehsmith** [ustaxcourt/ef-cms#8304]
## Pre-Conditions
ustaxcourt/ef-cms#6969 
ustaxcourt/ef-cms#8185 
ustaxcourt/ef-cms#8192

## Acceptance Criteria
- Refactor logic that serves a document across multiple cases, both court-issued and paper filings, to align with the business use cases. This is to improve the maintainability of existing code and make it easier to implement future consolidation stories.
- Better understand the less common business use case of deselecting some of the cases in the consolidated group to prevent them from being multi-docketed. This is as opposed to the much more common business use case of multi-docketing the entire consolidated group. 

## Notes
-  [x] Goal: Using Kibana, find out how many times multidocketing checkboxes have been unchecked
  - On a lead case while serving a court issued docket entry, find out how many times the docket entry has been served on some BUT NOT ALL cases in the consolidated group
- Goal: Refactor endpoints for serving court issued documents and paper filed documents to DRY up logic used for serving document across multiple cases in a consolidated group

## Tasks
- [x] Make useCaseHelper(s) for shared code
  - [x] fileAndServeCourtIssuedDocumentInteractor 
    - [x] Extract `fileDocumentOnOneCase` (https://github.com/flexion/ef-cms/pull/9686)
    - [x] Extract `stampDocument` (https://github.com/flexion/ef-cms/pull/9688)
    - [X] Refactor `closeCaseAndUpdateTrialSessionForEnteredAndServedDocuments` (https://github.com/flexion/ef-cms/pull/9695)
  - [x] serveCourtIssuedDocumentInteractor
    - [x] Refactor to use `stampDocumentUseCaseHelper` (https://github.com/flexion/ef-cms/pull/9689)
    - [x] Extract `serveDocumentOnOneCase` (https://github.com/flexion/ef-cms/pull/9694)
  - [x] ~~consolidate `serveCourtIssuedDocumentInteractor` and `fileAndServeCourtIssuedDocumentInteractor` into one~~ decided against this as there is just enough difference that it seems more readable to leave them separate
  - [x] serveExternallyFiledDocumentInteractor
    - [x] reuse `fileDocumentOnOneCase` - mob (https://github.com/flexion/ef-cms/pull/9701)
  - [x] addPaperFilingInteractor (https://github.com/flexion/ef-cms/pull/9706)
  - [x] `fileDocumentOnOneCase` investigate why DOCKET_SECTION is defaulted when work item is assigned to a user https://github.com/flexion/ef-cms/pull/9705
- [x] Standardize how docketNumbers are passed to different endpoints - Rosie https://github.com/flexion/ef-cms/pull/9702
- [x] Misc
  - [x] Lint warning fixes (https://github.com/flexion/ef-cms/pull/9691)
  - [x] Make an entity constant for served messages (https://github.com/flexion/ef-cms/pull/9696)
- [x] Regression testing
PR to test - https://github.com/ustaxcourt/ef-cms/pull/2784
FEEDBACK PR - https://github.com/ustaxcourt/ef-cms/pull/2822

## Story Definition of Ready (Created on 9/26/21)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Be immediately actionable for the development team
- [ ]  Be focused on users
- [ ]  Include a narrative. “As a ____, so that I can ______, I need _____.”
- [ ]  Include user-centric acceptance criteria that expresses the users' needs
- [ ]  Include user-centric testing scenarios
- [ ]  Include preconditions (story dependencies)
- [ ]  Not include design or technical solutions
- [ ]  Include non-standard testing, deployment, documentation, browser, mobile or security requirements, as well as any requirements for 3rd-party integrations

Process: Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [x] Automated test scripts have been written, including visual tests for newly added PDFs
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [ ] Reviewed by UX on a deployed environment.
 - [ ] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by rachelschneiderman** [comment]
@cholly75 

Research findings:

We wrote a script that retrieved all requests made to the following endpoints in the date range `7/17/2022-10/26/2022` and tracked the number of times the docket entry had been served on some BUT NOT ALL cases in the consolidated group, it's important to note that we excluded the situation where the docket entry was ONLY served on the lead case. (e.g. the docket entry was served on 2 out of 4 cases in a consolidated group).

`/case-documents/:subjectCaseDocketNumber/file-and-serve-court-issued-docket-entry` - 7 (2.7% of the time) instances of filing on some, but not all cases in a consolidated group out of 259 requests made to the endpoint. (48,000ish total requests to the endpoint).
`/case-documents/${docketNumber}/${docketEntryId}/serve-court-issued` - 0 (0% of the time) instances of filing on some, but not all cases in a consolidated group out of 1 requests made to the endpoint for a lead case (156 total requests to the endpoint).

**Comment by rachelschneiderman** [comment]
`serveCourtIssuedDocumentInteractor.ts`, `fileAndServeCourtIssuedDocumentInteractor.ts` are very similar and could be combined. We had some discussion regarding the benefits of combining and leaving them separate and ultimately decided to run the experiment of combining them to see if our concerns are alleviated once they are fully combined.

**Comment by rachelschneiderman** [comment]
The mob working on this story regression tested the code that was refactored by running through the test cases in ustaxcourt/ef-cms#6969 and ustaxcourt/ef-cms#8185. There were no failing tests!!!

**Comment by rachelschneiderman** [comment]
This story has been deployed to `ustc/test` and is ready to be tested. Waiting on the court to determine the testing procedure for this story as it doesn't have "traditional" test cases as it's a refactoring story.

**Comment by ttlenard** [comment]
## Test Cases - Consolidated Stories to Regression Test this Refactoring Story 

#### 1) 3569 - Save and Serve Court Issued documents
- [x] All tests from this story pass

Other Test Results:
**Need to verify Service works with Kibana Logs - Service failure to everyone while testing 11-16. Mike said he updated the bounced service email and service agent email yesterday. Probably caused this issue.**
- [x] Service works now

#### 2) 9513 - Save first then serve - from document preview pane - Court Issued Documents
- [x] All tests from this story pass

Other Test Results
**Need to verify Service works with Kibana Logs - Service failure to everyone while testing 11-16. Mike said he updated the bounced service email and service agent email yesterday. Probably caused this issue.**
- [x] Service works now


#### 3) 9585 - Save first, then serve - from document QC/In progress tabs - Court issued Documents
- [x] All tests from this story pass

Other Test Results:
**Need bug fix from 9718 to test this story**
- [x] Bug fix passes!


#### 4) 9501 - Unservable Court Filed Documents (Includes new coversheet - et al + docket/document No.)
- [x] All tests from this story pass

Other Test Results:
**Found Unrelated bug - Consolidated icons for these unservable document types are not appearing on the Document QC/Processed tab. Bug is ustaxcourt/ef-cms#8363**


#### 5) 2379 - Modify Case Caption/docket number on DAWSON Orders/Notices
- [x] All tests from this story pass


#### 6) 9493 - Court user files Paper Doc
- [ ] All tests from this story pass

Other Test results:
- [x] Getting 400 errors - The Case entity was invalid. {"filedBy":"'docketEntries[50].filedBy' is required"} and The Case entity was invalid. {"filedBy":"'docketEntries[43].filedBy' is required"} and he Case entity was invalid. {"filedBy":"'docketEntries[29].filedBy' is required"}

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/95c9acbf-7d9d-419b-a948-4b589df247e1)

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/54cc14c8-563d-4839-ad2d-9b82abd20375)

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/7e7ebc2f-1046-4c46-8e34-1283d1ee7cdd)

**Comment by ttlenard** [comment]
## Test Cases - Consolidated Stories to Regression Test the Refactoring Story (9681)

#### 1) 3569 - Save and Serve Court Issued documents
- [x] All tests from this story pass

#### 2) 9513 - Save first then serve - from document preview pane - Court Issued Documents
- [x] All tests from this story pass

#### 3) 9585 - Save first, then serve - from document QC/In progress tabs - Court issued Documents
- [x] All tests from this story pass

#### 4) 9501 - Unservable Court Filed Documents (Includes new coversheet - et al + docket/document No.)
- [x] All tests from this story pass
**Found Unrelated bug - Consolidated icons for these unservable document types are not appearing on the Document QC/Processed tab. I still need to enter this into Zenhub**

#### 5) 2379 - Modify Case Caption/docket number on DAWSON Orders/Notices
- [x] All tests from this story pass

#### 6) 9493 - Court user files Paper Doc
- [x] All tests from this story pass
